DOC-2524: Update: Callback Mode (Comments) (#3458)
tinymce/tinymce-docs: modules/ROOT/pages/comments-callback-mode.adoc

= Configuring the {pluginname} plugin in callback mode
:navtitle: Callback mode
:description: Information on configuring the Comments plugin in callback mode
:keywords: comments, commenting, tinycomments, callback
:pluginname: Comments
:plugincode: comments
:commentsMode: callback

> *Callback mode* is the default mode for the {pluginname} plugin. In the callback mode, callback functions are required to save user comments on a server. The {pluginname} functions (create, reply, edit, delete comment, delete all conversations, resolve, lookup, and fetch all conversations) are configured differently depending upon the server-side storage configuration.

Reviewer: ```suggestion
*Callback mode* is the default mode for the {pluginname} plugin. In the callback mode, callback functions are required to save user comments on a server. The {pluginname} functions (create, reply, edit, delete comment, delete all conversations, resolve, lookup, and fetch conversations) are configured differently depending upon the server-side storage configuration.
```

== How the {pluginname} plugin works in callback mode

All options accept functions incorporating `+done+` and `+fail+` callbacks as parameters. The function return type is not important, but all functions must call exactly one of these two callbacks: `+fail+` or `+done+`.

* The `+fail+` callback takes either a string or a JavaScript Error type.
* The `+done+` callback takes an argument specific to each function.

Most functions (create, reply, and edit) require an `+id+` identifying the *current author*.

[NOTE]
**Current author**: The {pluginname} plugin does not know the name of the current user. Determining the current user and storing the comment related to that user is required to be configured by the integrator.

> After a user comments (triggering `+tinycomments_create+` for the first comment or `+tinycomments_reply+` for subsequent comments), the {pluginname} plugin requests the updated conversation/s using both `+tinycomments_lookup+` and `+tinycomments_fetch+`, which contains the additional comment with the proper author.

Reviewer: I don't know exactly how comments use `tinycomments_fetch`. Although I notice that `tinycomments_fetch` is used instead of `tinycomments_lookup` if configured in the initial loading. 

This statement `the {pluginname} plugin requests the updated conversation/s using both `+tinycomments_lookup+` and `+tinycomments_fetch+`` needs to be confirmed.

> After a user comments (triggering `+tinycomments_create+` for the first comment or `+tinycomments_reply+` for subsequent comments), the {pluginname} plugin requests the updated conversation/s using both `+tinycomments_lookup+` and `+tinycomments_fetch+`, which contains the additional comment with the proper author.

Author: ```suggestion
When a user adds a comment (triggering `+tinycomments_create+` for the first comment or `+tinycomments_reply+` for replies), the {pluginname} plugin uses the `+tinycomments_lookup+` callback to retrieve the selected conversation. During the initial load, the `+tinycomments_fetch+` callback retrieves all existing annotated comments, including any additional comments with their respective authors.
```

> After a user comments (triggering `+tinycomments_create+` for the first comment or `+tinycomments_reply+` for subsequent comments), the {pluginname} plugin requests the updated conversation/s using both `+tinycomments_lookup+` and `+tinycomments_fetch+`, which contains the additional comment with the proper author.

Reviewer: ```suggestion
During the initial load, the {pluginname} uses `+tinycomments_fetch+` callback to retrieve the existing conversation including any additional comments with their respective authors . If not configured, the {pluginname} will fallback to `+tinycomments_lookup+`. 

When a user adds a comment, the {pluginname} plugin uses the `+tinycomments_lookup+` callback to retrieve the selected conversation. 
```

> After a user comments (triggering `+tinycomments_create+` for the first comment or `+tinycomments_reply+` for subsequent comments), the {pluginname} plugin requests the updated conversation/s using both `+tinycomments_lookup+` and `+tinycomments_fetch+`, which contains the additional comment with the proper author.

Reviewer: I added a suggestion based on your suggestion @kemister85

* xref:tinycomments_delete[`+tinycomments_delete+`]
* xref:tinycomments_delete_all[`+tinycomments_delete_all+`]
* xref:tinycomments_lookup[`+tinycomments_lookup+`]

> * xref:tinycomments_fetch[`+tinycomments_fetch+`]

Reviewer 2: This callback is optional as well. Current comments plugin users should not expect their implementation to break if they just upgrade.

CC: @Afraithe

Author: ```suggestion
```

include::partial$configuration/tinycomments_lookup.adoc[leveloffset=+1]

include::partial$configuration/tinycomments_fetch.adoc[leveloffset=+1]

=== Optional options

> * xref:tinycomments_resolve[`+tinycomments_resolve+`]
> 
> include::partial$configuration/tinycomments_resolve.adoc[leveloffset=+1]

Author: ```suggestion
* xref:tinycomments_resolve[`+tinycomments_resolve+`]
* xref:tinycomments_fetch[`+tinycomments_fetch+`]

include::partial$configuration/tinycomments_resolve.adoc[leveloffset=+1]

include::partial$configuration/tinycomments_fetch.adoc[leveloffset=+1]
```
